src/S3/S3MultiRegionClient.php

```diff
@@ -224,17 +227,31 @@ private function determineRegionMiddleware() {
                         $region = null;
                         if (isset($result['@metadata']['headers']['x-amz-bucket-region'])) {
                             $region = $result['@metadata']['headers']['x-amz-bucket-region'];
+                            $this->cache->set($cacheKey, $region);
                         } else {
-                            /** @var S3ClientInterface $client */
-                            $client = $this->getClientFromPool();
-                            $region = (yield $client->determineBucketRegionAsync(
+                            $region = $this->getDetermineBucketRegionGenerator(
                                 $command['Bucket']
-                            ));
+                            );
                         }
 
-                        $this->cache->set($cacheKey, $region);
                         $command['@region'] = $region;
                         yield $handler($command);
+                    } catch (AwsException $e) {
+                        if ($e->getAwsErrorCode() === 'AuthorizationHeaderMalformed') {
+                            $region = $this->determineBucketRegionFromExceptionBody(
+                                $e->getResponse()->getBody()
+                            );
+                            if (!empty($region)) {
+                                $this->cache->set($cacheKey, $region);
+
+                                $command['@region'] = $region;
+                                yield $handler($command);
+                            } else {
+                                throw $e;
+                            }
+                        } else {
```

Comment:
The generator will continue execution when looking for a potential next value to `yield` and would proceed to throw in all states past the `yield` invocation on L248 in doing so. This is different from the re-throw in `S3ClientTrait` because a the `return` blocks further execution for that scope.